api/desecapi/models.py

```diff
@@ -464,78 +298,39 @@ class RRset(models.Model, mixins.SetterMixin):
     class Meta:
         unique_together = (("domain", "subname", "type"),)
 
-    def __init__(self, *args, **kwargs):
-        self._dirties = set()
-        super().__init__(*args, **kwargs)
-
-    def setter_domain(self, val):
-        if val != self.domain:
-            self._dirties.add('domain')
-
-        return val
-
-    def setter_subname(self, val):
-        # On PUT, RRsetSerializer sends None, denoting the unchanged value
-        if val is None:
-            return self.subname
-
-        if val != self.subname:
-            self._dirties.add('subname')
-
-        return val
-
-    def setter_type(self, val):
-        if val != self.type:
-            self._dirties.add('type')
-
-        return val
+    @property
+    def name(self):
+        return '.'.join(filter(None, [self.subname, self.domain.name])) + '.'
 
-    def setter_ttl(self, val):
-        if val != self.ttl:
-            self._dirties.add('ttl')
+    def save(self, *args, **kwargs):
+        self.updated = timezone.now()
+        super().save(*args, **kwargs)
 
-        return val
+    def set_all_record_contents(self, record_contents):
+        """
+        Updates this RR set's resource records, discarding any old values.
 
-    def clean(self):
-        errors = {}
-        for field in (self._dirties & {'domain', 'subname', 'type'}):
-            errors[field] = ValidationError(
-                'You cannot change the `%s` field.' % field)
+        To do so, two large select queries and one query per changed (added or removed) resource record are needed.
 
-        if errors:
-            raise ValidationError(errors)
+        Changes are saved to the database immediately.
 
-    def get_dirties(self):
-        return self._dirties
+        :param record_contents: set of strings
+        """
+        # Remove RRs that we didn't see in the new list
+        removed_rrs = self.records.exclude(content__in=record_contents)  # one SELECT
+        for rr in removed_rrs:
+            rr.delete()  # one DELETE query
 
-    @property
-    def name(self):
-        return '.'.join(filter(None, [self.subname, self.domain.name])) + '.'
+        # Figure out which entries in record_contents have not changed
+        unchanged_rrs = self.records.filter(content__in=record_contents)  # one SELECT
+        unchanged_content = set([unchanged_rr.content for unchanged_rr in unchanged_rrs])
+        added_content = record_contents - unchanged_content
 
-    @transaction.atomic
-    def delete(self, *args, **kwargs):
-        self.domain.write_rrsets({self: []})
-        self._dirties = {}
+        for content in added_content:
+            RR(content=content, rrset=self).save()  # one INSERT query
```

Comment:
As you already discovered, `bulk_create()` does not trigger signals. With many records (think about importing dedyn.io), this probably makes a difference though. Maybe we should consider using `bulk_create()` and then triggering the signals after the fact. Here's an [example how to do that](https://stackoverflow.com/a/30632974/6867099) (there may be better ways).

Reply:
I believe the main benefit of using `bulk_create` is that no objects are created; however the signal includes the object instance. That's the technical issue here, but note that to us, this is actually not too important, as long as we know the `content` value.

My suggestion here is to first finish a working version of the change tracker and then consider improving performance. (This includes the risk that we have to scrape some work, but given this plan to improve performance I'd consider the risk to be acceptable.)

Reply:
jut